fcrepo-kernel-api/src/main/java/org/fcrepo/kernel/api/identifiers/FedoraID.java

```diff
@@ -50,24 +53,45 @@
 
     /**
      * Basic constructor.
-     * @param fullId The full identifier.
+     * @param fullId The full identifier or null if root.
      * @throws IllegalArgumentException If ID does not start with expected prefix.
      */
     public FedoraID(final String fullId) {
-        if (!fullId.startsWith(FEDORA_ID_PREFIX)) {
-            throw new IllegalArgumentException(String.format("ID must begin with %s", FEDORA_ID_PREFIX));
-        }
-        this.fullId = fullId;
+        this.fullId = ensurePrefix(fullId);
         if (!this.fullId.equals(FEDORA_ID_PREFIX)) {
             // Only strip trailing slashes the ID is more than the info:fedora/ prefix.
             this.fullId = this.fullId.replaceAll("/+$", "");
         }
         // Carry the path of the request for any exceptions.
-        this.externalPath = fullId.substring(FEDORA_ID_PREFIX.length());
+        this.externalPath = this.fullId.substring(FEDORA_ID_PREFIX.length());
 
         processIdentifier();
     }
 
+    /**
+     * Basic constructor for repository root ID.
+     */
+    private FedoraID() {
+        this("");
+    }
+
+    /**
+     * Static create method
+     * @param fullId The ID to use for the FedoraID.
+     * @return The FedoraID.
+     */
```

Comment:
Additionally, because I can't figure out how to leave comments on unchanged lines, I think it'd be nice if distinction between "fullId" and "resourceId" was documented more clearly (with id examples) in the javadocs for `getResourceId` and `getFullId`. Maybe in the `addTo` methods to? How are they related and why would I use one over the other?

Reply:
So the difference is really that `getResourceID` returns the ID to the actual resource and `getFullId` is the... full ID from the request.

So for a Memento, your full ID is `info:fedora/object/fcr:versions/2020040110101` and your resource ID is `info:fedora/object`.

However for some reason nonRdfSourceDescriptions need their resource Id to contain the `fcr:metadata` part. 

ie. Full ID : `info:fedora/object/fcr:metadata/fcr:versions/2020040110101`
Resource ID : `info:fedora/object/fcr:metadata`

I'm not sure why but I guess it has to do with how the file is stored with and read from the persistence layer.

Reply:
Yeah, that was my understanding, and that was a great explanation. Something like that should be in the javadoc.

Reply:
Added this to the top of the class, I hope that is okay.